File: src/main/java/com/braydenoneal/blang/parser/expression/builtin/ExportAllItemsBuiltin.java
package com.braydenoneal.blang.parser.expression.builtin;

import com.braydenoneal.blang.parser.Program;
import com.braydenoneal.blang.parser.RunException;
import com.braydenoneal.blang.parser.expression.Expression;
import com.braydenoneal.blang.parser.expression.ExpressionType;
import com.braydenoneal.blang.parser.expression.ExpressionTypes;
import com.braydenoneal.blang.parser.expression.value.*;
import com.mojang.serialization.Codec;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.LockableContainerBlockEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import java.util.List;

public record ExportAllItemsBuiltin(List<Expression> arguments) implements Expression {
    @Override
    public Value<?> evaluate(Program program) {
        Value<?> xValue = arguments.get(0).evaluate(program);
        Value<?> yValue = arguments.get(1).evaluate(program);
        Value<?> zValue = arguments.get(2).evaluate(program);
        Expression itemPredicateExpression = arguments.get(3);

        if (xValue instanceof IntegerValue x &&
                yValue instanceof IntegerValue y &&
                zValue instanceof IntegerValue z &&
                itemPredicateExpression instanceof FunctionValue itemPredicate
        ) {
            World world = program.context().entity().getWorld();

            if (world == null) {
                throw new RunException("World is null");
            }

            BlockPos entityPos = program.context().pos();
            BlockEntity exportEntity = world.getBlockEntity(new BlockPos(entityPos.getX() + x.value(), entityPos.getY() + y.value(), entityPos.getZ() + z.value()));

            if (exportEntity instanceof LockableContainerBlockEntity exportContainer) {
                List<LockableContainerBlockEntity> containers = program.context().entity().getConnectedContainers();

                for (LockableContainerBlockEntity container : containers) {
                    for (int i = 0; i < container.size(); i++) {
                        ItemStack stack = container.getStack(i);

                        program.newScope();
                        program.getScope().set(itemPredicate.value().arguments().getFirst(), new ItemValue(stack.getItem()));
                        Value<?> predicateResult = itemPredicate.call(program);
                        program.endScope();

                        if (predicateResult instanceof BooleanValue booleanValue && booleanValue.value()) {
                            for (int j = 0; j < exportContainer.size(); j++) {
                                ItemStack exportStack = exportContainer.getStack(j);

                                if (exportStack.isOf(Items.AIR)) {
                                    container.removeStack(i);
                                    exportContainer.setStack(j, stack);
                                    break;
                                }

                                if (exportStack.isOf(stack.getItem())) {
                                    int move = Math.min(stack.getCount(), exportStack.getMaxCount() - exportStack.getCount());

                                    stack.decrement(move);
                                    exportStack.increment(move);

                                    if (stack.isEmpty()) {
                                        container.removeStack(i);
                                        exportContainer.setStack(j, exportStack);
                                        break;
                                    } else {
                                        container.setStack(i, stack);
                                        exportContainer.setStack(j, exportStack);
                                    }
                                }
                            }
                        }
                    }
                }

                return Null.value();
            }
        }

        throw new RunException("Invalid arguments");
    }

    public static final MapCodec<ExportAllItemsBuiltin> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
            Codec.list(Expression.CODEC).fieldOf("arguments").forGetter(ExportAllItemsBuiltin::arguments)
    ).apply(instance, ExportAllItemsBuiltin::new));

    @Override
    public ExpressionType<?> getType() {
        return ExpressionTypes.EXPORT_ALL_ITEMS_BUILTIN;
    }
}

File: src/main/java/com/braydenoneal/blang/parser/expression/builtin/ExportAllItemsBuiltin.kt
package com.braydenoneal.blang.parser.expression.builtin

import com.braydenoneal.blang.parser.Program
import com.braydenoneal.blang.parser.RunException
import com.braydenoneal.blang.parser.expression.Arguments
import com.braydenoneal.blang.parser.expression.Expression
import com.braydenoneal.blang.parser.expression.ExpressionType
import com.braydenoneal.blang.parser.expression.ExpressionTypes
import com.braydenoneal.blang.parser.expression.value.BooleanValue
import com.braydenoneal.blang.parser.expression.value.ItemValue
import com.braydenoneal.blang.parser.expression.value.Null
import com.braydenoneal.blang.parser.expression.value.Value
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.block.entity.LockableContainerBlockEntity
import net.minecraft.item.Items
import net.minecraft.util.math.BlockPos
import java.util.Map
import kotlin.math.min


data class ExportAllItemsBuiltin(val arguments: Arguments) : Expression {
    override fun evaluate(program: Program): Value<*> {
        val x = arguments.integerValue(program, "x", 0).value()
        val y = arguments.integerValue(program, "y", 1).value()
        val z = arguments.integerValue(program, "z", 2).value()
        val itemPredicate = arguments.functionValue(program, "itemPredicate", 3)

        val world = program.context().entity!!.getWorld()

        if (world == null) {
            throw RunException("World is null")
        }

        val entityPos = program.context().pos
        val exportEntity = world.getBlockEntity(BlockPos(entityPos.x + x, entityPos.y + y, entityPos.z + z))

        if (exportEntity !is LockableContainerBlockEntity) {
            throw RunException("Block at position is not a container")
        }

        val containers = program.context().entity!!.getConnectedContainers()

        for (container in containers) {
            for (slot in 0..<container.size()) {
                val stack = container.getStack(slot)

                val predicateArguments = Arguments(mutableListOf(ItemValue(stack.item)), Map.of())
                val predicateResult = itemPredicate.call(program, predicateArguments)

                if (predicateResult !is BooleanValue) {
                    throw RunException("itemPredicate is not a predicate")
                }

                if (!predicateResult.value()) {
                    continue
                }

                for (exportSlot in 0..<exportEntity.size()) {
                    val exportStack = exportEntity.getStack(exportSlot)

                    if (exportStack.isOf(Items.AIR)) {
                        container.removeStack(slot)
                        exportEntity.setStack(exportSlot, stack)
                        break
                    }

                    if (!exportStack.isOf(stack.item)) {
                        continue
                    }

                    val move = min(stack.count, exportStack.maxCount - exportStack.count)

                    stack.decrement(move)
                    exportStack.increment(move)

                    if (stack.isEmpty) {
                        container.removeStack(slot)
                        exportEntity.setStack(exportSlot, exportStack)
                        break
                    }

                    container.setStack(slot, stack)
                    exportEntity.setStack(exportSlot, exportStack)
                }
            }
        }

        return Null.VALUE
    }

    override val type: ExpressionType<*> get() = ExpressionTypes.EXPORT_ALL_ITEMS_BUILTIN

    companion object {
        val CODEC: MapCodec<ExportAllItemsBuiltin> = RecordCodecBuilder.mapCodec { instance ->
            instance.group(
                Arguments.CODEC.fieldOf("arguments").forGetter(ExportAllItemsBuiltin::arguments)
            ).apply(instance, ::ExportAllItemsBuiltin)
        }
    }
}
